operation/web.go

```diff
@@ -0,0 +1,118 @@
+package operation
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const opidMetadataKey = "opid"
+
+// UnaryClientInterceptor adds the operation id from the current context (if any) to the
+// outgoing unary RPC metadata.
+func UnaryClientInterceptor(
+	ctx context.Context,
+	method string,
+	req, reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	if op := Get(ctx); op != nil && op.ID.String() != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, opidMetadataKey, op.ID.String())
+	}
+	return invoker(ctx, method, req, reply, cc, opts...)
+}
+
+// StreamClientInterceptor adds the operation id from the current context (if any) to the
+// outgoing streaming RPC metadata.
+func StreamClientInterceptor(
+	ctx context.Context,
+	desc *grpc.StreamDesc,
+	cc *grpc.ClientConn,
+	method string,
+	streamer grpc.Streamer,
+	opts ...grpc.CallOption,
+) (grpc.ClientStream, error) {
+	if op := Get(ctx); op != nil && op.ID.String() != "" {
+		ctx = metadata.AppendToOutgoingContext(ctx, opidMetadataKey, op.ID.String())
+	}
+	return streamer(ctx, desc, cc, method, opts...)
+}
+
+// UnaryServerInterceptor creates a new operation in the current context before passing
+// it to the unary response handler. If the incoming RPC metadata contains an operation
+// ID, the new operation will have the same ID.
+func (m *Manager) UnaryServerInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (interface{}, error) {
+	meta, ok := metadata.FromIncomingContext(ctx)
```

Comment:
broader question here - should we be erroring in this interceptor? I feel like a malformed metadata isn't that big of a deal and we can continue with the request (and just make our own op id)

Reply:
https://github.com/viamrobotics/rdk/pull/1501/commits/7a71b89f8bf83baede64855e46828701320eb040